File: src/main/java/com/github/dr/rwserver/util/game/Events.java
package com.github.dr.rwserver.util.game;

import com.github.dr.rwserver.func.Cons;
import com.github.dr.rwserver.struct.ObjectMap;
import com.github.dr.rwserver.struct.Seq;

/**
 * @author Dr
 */
@SuppressWarnings("unchecked")
public class Events {
    private static final ObjectMap<Object, Seq<Cons<?>>> EVENTS = new ObjectMap<>();

    public static <T> void on(Class<T> type, com.github.dr.rwserver.func.Cons<T> listener) {
        EVENTS.get(type, Seq::new).add(listener);
    }

    public static void on(Object type, Runnable listener) {
        EVENTS.get(type, Seq::new).add(e -> listener.run());
    }

    public static <T> void remove(Class<T> type, com.github.dr.rwserver.func.Cons<T> listener) {
        EVENTS.get(type, Seq::new).remove(listener);
    }

    public static <T> void fire(T type) {
        fire(type.getClass(), type);
    }

    public static <T> void fire(Class<?> type1, T type) {
        if(EVENTS.get(type) != null) {
            EVENTS.get(type).each(e -> ((com.github.dr.rwserver.func.Cons<T>)e).get(type));
        }
        if(EVENTS.get(type1) != null) {
            EVENTS.get(type1).each(e -> ((com.github.dr.rwserver.func.Cons<T>)e).get(type));
        }
    }

}

File: Server-Core/src/main/java/net/rwhps/server/util/game/Events.kt
/*
 * Copyright 2020-2024 RW-HPS Team and contributors.
 *
 * 此源代码的使用受 GNU AFFERO GENERAL PUBLIC LICENSE version 3 许可证的约束, 可以在以下链接找到该许可证.
 * Use of this source code is governed by the GNU AGPLv3 license that can be found through the following link.
 *
 * https://github.com/RW-HPS/RW-HPS/blob/master/LICENSE
 */
package net.rwhps.server.util.game

import kotlinx.coroutines.*
import net.rwhps.server.func.Control
import net.rwhps.server.game.event.core.AbstractEventCore
import net.rwhps.server.struct.map.ObjectMap
import net.rwhps.server.struct.list.Seq
import net.rwhps.server.util.annotations.core.EventOnlyRead
import net.rwhps.server.util.inline.ifNull

/**
 * @author Dr (dr@der.kim)
 */
class Events {
    private val eventData = ObjectMap<Any, Seq<(Any) -> Any?>>()
    private val eventMonitorScope = CoroutineScope(SupervisorJob() + Dispatchers.Default)

    fun <T> addEvent(type: Class<T>, listener: (T) -> Unit) {
        addEvent(false, type, listener)
    }

    @Suppress("UNCHECKED_CAST")
    fun <T> addEvent(async: Boolean, type: Class<T>, listener: (T) -> Unit) {
        if (async || type.getAnnotation(EventOnlyRead::class.java) != null) {
            eventData[type, { Seq() }].add { value ->
                return@add eventMonitorScope.launch {
                    listener(value as T)
                }
            }
        } else {
            eventData[type, { Seq() }].add { value ->
                listener(value as T)
                return@add null
            }
        }
    }

    fun addEvent(type: Any, listener: () -> Unit) {
        eventData[type, { Seq() }].add { listener() }
    }

    fun <T> remove(type: Class<T>) {
        eventData[type, { Seq() }].clear()
    }

    fun <T> fire(type1: Class<*>, type: T) {
        val async = Seq<Job>()

        val eventType = eventData[type as Any]
        eventType?.eachAll { e: (Any) -> Any? -> runAsync(async, e, type) }

        val eventType1 = eventData[type1]
        eventType1?.eachAll { e: (Any) -> Any? -> runAsync(async, e, type) }

        runBlocking {
            async.forEach {
                it.join()
            }
        }
    }

    private fun <T> runAsync(async: Seq<Job>, e: (T) -> Any?, type: T) {
        if (type is AbstractEventCore) {
            if (type.status() == Control.EventNext.CONTINUE) {
                e(type).ifNull({
                    if (it is Job) {
                        async.add(it)
                    }
                })
            }
        }
    }
}